pkg/registry/core/service/storage/rest.go

```diff
@@ -170,25 +171,28 @@ func (rs *REST) Create(ctx context.Context, obj runtime.Object, createValidation
 	}()
 
 	var err error
-	if service.Spec.Type != api.ServiceTypeExternalName {
-		if releaseServiceIP, err = initClusterIP(service, rs.serviceIPs); err != nil {
-			return nil, err
+	var nodePortOp *portallocator.PortAllocationOperation
+	if !dryrun.IsDryRun(options.DryRun) {
+		if service.Spec.Type != api.ServiceTypeExternalName {
+			if releaseServiceIP, err = initClusterIP(service, rs.serviceIPs); err != nil {
+				return nil, err
+			}
 		}
-	}
 
-	nodePortOp := portallocator.StartOperation(rs.serviceNodePorts)
-	defer nodePortOp.Finish()
+		nodePortOp = portallocator.StartOperation(rs.serviceNodePorts)
```

Comment:
Did you consider adding a dry-run concept to this operation thingy?

I think the extra if statements make this code a lot harder to follow. I'm not sure what the best way to fix that is. Maybe provide separate dry run / non dry run functions, and call the right one? Or make the nodePortOp take a dry run flag.